File: src/com/douma/笔试代码/zoom/_20220810/_1_TreeTotalWeight.py
# 抖码算法，让算法学习变得简单有趣
# 作者：老汤


class TreeNode:
    def __init__(self, left=None, right=None, is_red=False):
        self.left = left
        self.right = right
        # true 表示红节点，false 表示蓝色节点
        self.is_red = is_red
        # self.red_num：用于记录到这个节点为止，路径上红色节点的个数
        # self.blue_num：用于记录到这个节点为止，路径上蓝色节点的个数
        if is_red:
            self.red_num, self.blue_num = 1, 0
        else:
            self.red_num, self.blue_num = 0, 1

    def get_weight(self):
        return abs(self.red_num - self.blue_num)

"""
* 求树的权值
*
* 告诉你一棵树的节点分为蓝色和红色，一个节点的权值定义为：
* 从根节点到该节点路径上的红蓝节点个数之差的绝对值
*
* 求整棵树的权值之和
"""
def total_weight(root):
    res = 0
    if root is None:
        return res

    res += root.get_weight()

    def dfs(node):
        if node is None:
            return

        nonlocal res
        if node.left:
            node.left.red_num += node.red_num
            node.left.blue_num += node.blue_num
            res += node.left.get_weight()
            dfs(node.left)
        if node.right:
            node.right.red_num += node.red_num
            node.right.blue_num += node.blue_num
            res += node.right.get_weight()
            dfs(node.right)

    dfs(root)
    return res


"""
构造一颗下面的树
                    红(1)
                  /    \
                蓝(0)   红(2)
              /   \
           红(1)  蓝(1)
"""
root = TreeNode(is_red=True)

node1 = TreeNode(is_red=False)
root.left = node1

node2 = TreeNode(is_red=True)
root.right = node2

node3 = TreeNode(is_red=True)
node1.left = node3

node4 = TreeNode(is_red=False)
node1.right = node4

print(total_weight(root))

File: src/com/douma/笔试代码/zoom/_20220810/_1_TreeTotalWeight.java
package com.douma.笔试代码.zoom._20220810;

/**
 * 抖码算法，让算法学习变的简单有趣
 *
 * @作者 : 老汤
 */

class TreeNode {
    TreeNode left;
    TreeNode right;
    boolean isRed; // true 表示红节点，false 表示蓝色节点
    // 用于记录到这个节点为止，路径上红色节点的个数
    int redNum;
    // 用于记录到这个节点为止，路径上蓝色节点的个数
    int blueNum;

    TreeNode(boolean isRed) {
        this.isRed = isRed;
        this.redNum = isRed ? 1 : 0;
        this.blueNum = isRed ? 0 : 1;
    }

    int getWeight() {
        return Math.abs(redNum - blueNum);
    }
}

public class _1_TreeTotalWeight {
    /**
     * 求树的权值
     *
     * 告诉你一棵树的节点分为蓝色和红色，一个节点的权值定义为：
     * 从根节点到该节点路径上的红蓝节点个数之差的绝对值
     *
     * 求整棵树的权值之和
     */

    int res = 0;

    public int totalWeight(TreeNode root) {
        if (root == null) return res;

        res += root.getWeight();

        dfs(root);

        return res;
    }

    private void dfs(TreeNode node) {
        if (node == null) return;

        if (node.left != null) {
            node.left.redNum += node.redNum;
            node.left.blueNum += node.blueNum;
            res += node.left.getWeight();
            dfs(node.left);
        }
        if (node.left != null) {
            node.right.redNum += node.redNum;
            node.right.blueNum += node.blueNum;
            res += node.right.getWeight();
            dfs(node.right);
        }
    }

    public static void main(String[] args) {
        /**
         构造一颗下面的树
                    红(1)
                  /    \
                蓝(0)   红(2)
              /   \
           红(1)  蓝(1)
         */
        TreeNode root = new TreeNode(true);

        TreeNode node1 = new TreeNode(false);
        root.left = node1;

        TreeNode node2 = new TreeNode(true);
        root.right = node2;

        TreeNode node3 = new TreeNode(true);
        node1.left = node3;

        TreeNode node4 = new TreeNode(false);
        node1.right = node4;

        System.out.println(new _1_TreeTotalWeight().totalWeight(root));
    }
}
